File: introduction/insertionSort.cpp
#include <iostream>

void insertionSort(int*, int);
void print(int*, int);

int main(){
    // insertion sort
    int A[] = {6,3,8,2,1,9,8,0,7};
    int *p_A = A;
    int size = sizeof(A)/sizeof(*A);
    
    print(p_A, size);
    insertionSort(p_A, size);
    std::cout << "Sorting ascending: " << "\n";
    print(p_A, size);
    return 0;
}

void insertionSort(int *A, int size){
    
    for(int j = 1; j < size; ++j){
        int key = A[j];
        int i = j - 1;
        while(i >= 0 && A[i] > key){
            A[i + 1] = A[i];
            i = i - 1;
        }
        A[i + 1] = key;
    }
}

void print(int* A, int size){
    for(int i = 0; i < size;++i){
        std::cout << *(A+i) << " ";
    }
    std::cout << "\n";
}

File: sorting-algorithms/insertionSort.js
const numbers = require('./unsortedNumbers');
/**
 *      INSERTION SORT
 *
 *          Builds up the sort by gradually creating a larger left half which
 *          is always sorted
 * */
let unsortedNums = numbers.unsortedNumbers;

// Time complexity: O(n^2)
function insertionSortv2(list) {
    for (let i = 1; i < list.length; i++) {
        let pos = 0;
        // looking for new position for list[i], in left ordered window
        for (let j = 0; j < i; j++) {
            if(list[i] > list[j]) pos = j+1;
        }
        // insertig the list[i] element in pos index within left half window
        // recolocating every element in left window on place forward, up to the i position
        let actualVal = 0;
        let pastValue = list[pos];
        list[pos] = list[i];
        for(let j=pos + 1; j <= i; j++) {
            actualVal = list[j];
            list[j] = pastValue;
            pastValue = actualVal;
        }
        //console.log(list)
    }
    return list;
}

// swapping j and j-1 until list[j] > list[j-1]
function insertionSortv3(list) {
    for (let i = 1; i < list.length; i++)
        for (let j = i; j > 0 && list[j] < list[j-1]; j--)
            [list[j], list[j-1]] = [list[j-1], list[j]]
    return list;
}

// assigning j index value to j+1 until we get proper j position for current value
function insertionSort(list) {
    for (let i = 1; i < list.length; i++) {
        let currentValue = list[i];
        let j;
        for (j = i - 1; j >= 0 && list[j] > currentValue ; j--) {
            list[j+1] = list[j];
        }
        list[j+1] = currentValue;
    }
    return list;
}

console.log(insertionSort(unsortedNums));
console.log(insertionSort([5, 3, 7, 2, 9, 1, 6, 4, 9, 10, 8]));
console.log(insertionSort([5, 3, 4, 1, 2]));
